### PR TITLE
[AArch64] Add tests for redundant csel instructions. NFC

### PR DESCRIPTION
**Pull request by citymarina**

### CODE REVIEW

**Review by davemgreen — COMMENTED** [review]
Hello - The mirs test look OK. The llvm-ir has some dead instructions that could be removed.

Can you give some details on the optimization they are hoping to test?

**Comment by citymarina** [comment]
1613781 is the followup that I plan to post after this.

Regarding the dead instructions, are you referring to the assignments to `%cmp1` and `%and.i`? I tried to remove these, but then the `csel` that I want to test is no longer produced (instead there is a `mvn` or `orn`).

I was suggested to include the `.ll` file as more of an end-to-end test, to guard against the possibility that maybe the bad `csel` might return at a later stage, which a narrow `-run-pass` test case wouldn't catch. However I am now starting to think that maybe this test is not enough of a controlled environment. If the test depends on having dead instructions, then that seems prone to breakage from random future changes. Do you have any advice on how to proceed?

**Comment by davemgreen** [comment]
Hi - yeah having the ll test is useful, I see the csels come from ifcvt. It can be nice to makes sure that optimizations happen early to avoid issues like this, but if they are being produced so late then the peephole sounds OK to me.

The test might not be super reliable as code changes over time, but if the store/call doesn't work then it is probably OK so long as we have the mir tests too.

File: llvm/test/CodeGen/AArch64/peephole-csel.ll [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+; CHECK-NEXT:    ret
+entry:
+  br i1 %0, label %then, label %exit
+
+then:                                             ; preds = %entry
+  %cmp1 = icmp eq i32 0, 0
+  br i1 %cmp, label %true, label %exit
+
+true:                                             ; preds = %then
+  %and.i = lshr i64 0, 0
```
[the remark this answers is not in the document]
Comment (citymarina) [reply llvm/test/CodeGen/AArch64/peephole-csel.ll:22]: It seems that any change I make is enough to disturb the test case, with one exception: I could replace the two dead instructions with `call void @llvm.donothing()`. What do you think?

**Review by davemgreen — APPROVED** [review]
That's a new one to me. This sounds OK so long as we have the mir tests too. LGTM

**Pull request merged**